tools/fv_treat_da_inc.F90

```diff
@@ -461,6 +472,183 @@ end subroutine apply_inc_on_3d_scalar
+                   ua_inc(isc_in-1,j,k) = ua_inc(isc_in,j,k)
+                   va_inc(isc_in-1,j,k) = va_inc(isc_in,j,k)
+                enddo
+             enddo
+          endif
+          if ( iec_in == Atm%npx ) then
+             do k = 1,npz_in
+                do j = jsc_in,jec_in
+                   ua_inc(iec_in+1,j,k) = ua_inc(iec_in,j,k)
+                   va_inc(iec_in+1,j,k) = va_inc(iec_in,j,k)
+                enddo
+             enddo
+          endif
+          if ( jsc_in == 1 ) then
+             do k = 1,npz_in
+                do i = isc_in,iec_in
+                   ua_inc(i,jsc_in-1,k) = ua_inc(i,jsc_in,k)
+                   va_inc(i,jsc_in-1,k) = va_inc(i,jsc_in,k)
+                enddo
+             enddo
+          endif
+          if ( jec_in == Atm%npy ) then
+             do k = 1,npz_in
+                do i = isc_in,iec_in
+                   ua_inc(i,jec_in+1,k) = ua_inc(i,jec_in,k)
+                   va_inc(i,jec_in+1,k) = va_inc(i,jec_in,k)
+                enddo
+             enddo
+          endif
+
+          ! corners
+          do k = 1,npz_in
+             if ( isc_in == 1 .and. jsc_in == 1 ) then
+                ua_inc(isc_in-1,jsc_in-1,k) = ua_inc(isc_in,jsc_in,k)
+                va_inc(isc_in-1,jsc_in-1,k) = va_inc(isc_in,jsc_in,k)
+             elseif ( isc_in == 1 .and. jec_in == Atm%npy ) then
+                ua_inc(isc_in-1,jec_in+1,k) = ua_inc(isc_in,jec_in,k)
+                va_inc(isc_in-1,jec_in+1,k) = va_inc(isc_in,jec_in,k)
+             elseif ( iec_in == Atm%npx .and. jsc_in == 1 ) then
+                ua_inc(iec_in+1,jsc_in-1,k) = ua_inc(iec_in,jec_in,k)
+                va_inc(iec_in+1,jsc_in-1,k) = va_inc(iec_in,jec_in,k)
+             elseif ( iec_in == Atm%npx .and. jec_in == Atm%npy ) then
+                ua_inc(iec_in+1,jec_in+1,k) = ua_inc(iec_in,jec_in,k)
+                va_inc(iec_in+1,jec_in+1,k) = va_inc(iec_in,jec_in,k)
+             endif
+          enddo
+       endif
+
+       ! Apply A-grid wind increments to D-grid winds
+       call update_dwinds_phys(isc_in, iec_in, jsc_in, jec_in, is_in, ie_in, js_in, je_in, &
+                               1., ua_inc, va_inc, u, v, &
+                               Atm%gridstruct, Atm%npx, Atm%npy, npz_in, fv_domain)
+    endif
+
+    ! Remaining increments
+    ! --------------------
+    
+    ! Get tracer indices
+    sphum   = get_tracer_index(MODEL_ATMOS, 'sphum')
+    liq_wat = get_tracer_index(MODEL_ATMOS, 'liq_wat')
+#ifdef MULTI_GASES
+    spo     = get_tracer_index(MODEL_ATMOS, 'spo')
+    spo2    = get_tracer_index(MODEL_ATMOS, 'spo2')
+    spo3    = get_tracer_index(MODEL_ATMOS, 'spo3')
+#else
+    o3mr    = get_tracer_index(MODEL_ATMOS, 'o3mr')
+#endif
+
+    ! Apply increments
+    do k = 1,npz_in
+       do j = jsc_in,jec_in
+          do i = isc_in,iec_in
+             pt(i,j,k)   = pt(i,j,k)   + increment_data%temp_inc(i,j,k)
+             delp(i,j,k) = delp(i,j,k) + increment_data%delp_inc(i,j,k)
```

Comment:
Do make sure that the delp increment is consistent with the tracer increments, as tracers are defined as the total-mass specific ratio and delp the mass of dry air, water vapor, and all condensates.

Reply:
@lharris4 You mean in JEDI, right? That's a good point, because while I have to look at how delp is defined in JEDI, I do know that the delz increment is currently defined hydrostatically as the difference between the hydrostatic delz and the beginning and end of the cycle, and that hydrostatic relationship is determined using a virtual temperature that only accounts for dry air and water vapor, but no condensates. It's something I've been meaning to address.

Reply:
@DavidNew here I am referring to `delp` and the adjustment of tracer masses when `delp` is incremented, as is done in `fv_update_phys()`. This is because each tracer is defined as the specific ratio of tracer mass and total air mass (`delp`) in a grid cell, and when `delp` is updated the denominator is changed; so to ensure tracer mass is conserved the tracers need to be adjusted to account for the change in `delp`

Reply:
Got it. I'll make a note to look into verifying this in JEDI. Thanks

Reply:
@lharris4 sorry to take you back a long time, we have just been discussing something at EMC. Is the process you're referring to done using the remap methods normally? I wonder if this is similar to what is needed when starting the model from a cold start. Is it true to say (in some sense) that once an increment is added to the state of the model you have a 'cooler start' because the incrementing of pressure moves all the model layers around?

Reply:
Hi, Dan. We perform the adjustments when doing the intialization, as in `external_ic_mod::nggps_ic()`. I would presume that if you are doing DA cycling no adjustments are necessary because the cycling is done on outputs from the same driving model. I am simply making a cautionary note to make sure that the DA cycling system isn't transforming the increments into something unexpected. One possibility is that if inconsistent increments are applied to both delp and tracers then there may be an unphysical change in the dry-air mass.

Reply:
Thanks for the response Lucas. That's a great point about the need to balance how the tracers are defined. Are there routines for doing that in FV3?

Is it not also true that the levels of the model are changed if DELP is changed? And thus that the variables stored on those levels, temperature, winds etc also need to be interpolated vertically to the new levels? Perhaps there is a difference between when running hydrostatically and non-hydrostatically here to consider as well.

Reply:
Hi, Dan. This is done within this block of code here:

[https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/main/model/fv_update_phys.F90#L328-L367](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/main/model/fv_update_phys.F90#L328-L367)

For your purposes, `adj_mass_vmr = 0`.

Please let me know if you need any further help.

Thanks
Lucas